src/ast/passes/codegen_llvm.h

```diff
@@ -280,19 +274,74 @@ class CodegenLLVM : public Visitor {
 
   std::map<std::string, AllocaInst *> variables_;
 
-  // NB: ensure all IDs are saved/restored in create_reset_ids()
-  int printf_id_ = 0;
-  int mapped_printf_id_ = 0;
-  int time_id_ = 0;
-  int cat_id_ = 0;
-  int strftime_id_ = 0;
-  uint64_t join_id_ = 0;
-  int system_id_ = 0;
-  int non_map_print_id_ = 0;
-  uint64_t watchpoint_id_ = 0;
-  int cgroup_path_id_ = 0;
-  int skb_output_id_ = 0;
-  int str_id_ = 0;
+#define ADD_ASYNC_ID(ID)                                                       \
+  int ID = 0;                                                                  \
+  int get_##ID()                                                               \
+  {                                                                            \
+    return ID++;                                                               \
+  }
+
+// For 'create_reset_ids' return a lambda that has captured-by-value
+// CodegenLLVM's async id state. Running the returned lambda will restore
+// `CodegenLLVM`s async id state back to when this function was first called.
+#define CREATE_ASYNC_IDS(                                                      \
+    ID1, ID2, ID3, ID4, ID5, ID6, ID7, ID8, ID9, ID10, ID11, ID12)             \
```

Comment:
I think there is a better way to do this with boost but not sure it's worth adding.

Reply:
I think it might be cleaner to use an x-macro: https://en.wikipedia.org/wiki/X_macro 

New IDs can be added with a single line